ema-oracle/src/types.rs

```diff
@@ -0,0 +1,207 @@
+// This file is part of pallet-ema-oracle.
+
+// Copyright (C) 2022  Intergalactic, Limited (GIB).
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::sp_runtime::RuntimeDebug;
+use hydra_dx_math::ema::{
+    balance_weighted_average, exp_smoothing, price_weighted_average, volume_weighted_average, EmaPrice,
+};
+use hydra_dx_math::types::Fraction;
+use hydradx_traits::{AggregatedEntry, Volume};
+use scale_info::TypeInfo;
+use sp_arithmetic::traits::{AtLeast32BitUnsigned, SaturatedConversion, UniqueSaturatedInto};
+
+pub use hydradx_traits::{OraclePeriod, Source};
+
+use sp_std::prelude::*;
+
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+
+pub type AssetId = u32;
+pub type Balance = u128;
+/// A price is a tuple of two `u128`s representing the numerator and denominator of a rational number.
+pub type Price = EmaPrice;
+
+/// A type representing data produced by a trade or liquidity event. Timestamped to the block where
+/// it was created.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(RuntimeDebug, Encode, Decode, Clone, PartialEq, Eq, Default, TypeInfo, MaxEncodedLen)]
+pub struct OracleEntry<BlockNumber> {
+    pub price: Price,
+    pub volume: Volume<Balance>,
+    pub liquidity: Balance,
+    pub timestamp: BlockNumber,
+}
+
+impl<BlockNumber> OracleEntry<BlockNumber>
+where
+    BlockNumber: AtLeast32BitUnsigned + Copy + UniqueSaturatedInto<u64>,
+{
+    /// Convert the `OracleEntry` into an `AggregatedEntry` for consumption. Determines the age by
+    /// subtracting `initialized` from the timestamp.
+    pub fn into_aggregated(self, initialized: BlockNumber) -> AggregatedEntry<Balance, BlockNumber, Price> {
+        AggregatedEntry {
+            price: self.price,
+            volume: self.volume,
+            liquidity: self.liquidity,
+            oracle_age: self.timestamp.saturating_sub(initialized),
+        }
+    }
+
+    /// Return an inverted version of the entry where the meaning of assets a and b are inverted.
+    /// So the price of a/b become the price b/a and the volume switches correspondingly.
+    pub fn inverted(&self) -> Self {
```

Comment:
not that important, although I'd think that you use the inverted entry and there would not be a need to keep the previous one and therefore you could consider:

```suggestion
    pub fn inverted(self) -> Self {
```